README
NAME
    Dancer2::Plugin::TemplateFlute - Dancer2 form handler for
    Template::Flute template engine

VERSION
    Version 0.201

SYNOPSIS
    Display template with checkout form:

        get '/checkout' => sub {
            my $form;

            $form = form( name => 'checkout', source => 'session' );
        
            template 'checkout', { form => $form };
        };

    Retrieve form input from checkout form body:

        post '/checkout' => sub {
            my ($form, $values);

            $form = form( name => 'checkout', source => 'body' );
            $values = $form->values;
        };

    Reset form after completion to prevent old data from showing up on new
    form:

        form('checkout')->reset;

    If you have multiple forms then just pass the form token as an array
    reference:

        get '/cart' => sub {
            my ( $cart_form, $inquiry_form );

            # 'source' defaults to 'session' if not provided
            $cart_form    = form( name => 'cart' );

            # equivalent to form( name => 'inquiry' )
            $inquiry_form = form( 'inquiry' );
        
            template 'checkout', { form => [ $cart_form, $inquiry_form ] };
        };

KEYWORDS
  form <$name|%params>
    The following %params are recognised:

   name
    The name of the form. Defaults to 'main'.

   values
        my $form = form( 'main', values => $params );

    The form parameters as a Hash::MultiValue object or something that can
    be coerced into one.

    Instead of "values" you can also use "source" to set initial values.

   source
        my $form = form( name => 'main', source => 'body' );

    The following values are valid:

    body
        This sets the form values to the request body parameters
        Dancer2::Core::Request::body_parameters.

    query
        This sets the form values to the request query parameters
        Dancer2::Core::Request::query_parameters.

    parameters
        This sets the form values to the combined body and request query
        parameters Dancer2::Core::Request::parameters.

    session
        Reads in values from the session. This is the default if no "source"
        or "parameters" are specified.

    NOTE: if both "source" and "values" are supplied then values is ignored.

    See Dancer2::Plugin::TemplateFlute::Form for details of other parameters
    that can be used.

DESCRIPTION
    "Dancer2::Plugin::TemplateFlute" is used for forms with the
    Dancer2::Template::TemplateFlute templating engine.

    Form fields, values and errors are stored into and loaded from the
    session key "form".

AUTHORS
    Original Dancer plugin by:

    Stefan Hornburg (Racke), "<racke at linuxia.de>"

    Initial port to Dancer2 by:

    Evan Brown (evanernest), "evan at bottlenose-wine.com"

    Rehacking to Dancer2's plugin2 and general rework:

    Peter Mottram (SysPete), "peter at sysnix.com"

BUGS
    Please report any bugs or feature requests via GitHub issues:
    <https://github.com/interchange/Dancer2-Plugin-TemplateFlute/issues>.

    We will be notified, and then you'll automatically be notified of
    progress on your bug as we make changes.

ACKNOWLEDGEMENTS
LICENSE AND COPYRIGHT
    Copyright 2011-2016 Stefan Hornburg (Racke).

    Copyright 2015-1016 Evan Brown.

    Copyright 2016 Peter Mottram (SysPete).

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Perl module Dancer2::Plugin::TemplateFlute

0.201 Mon  6 Jun 18:41:52 CEST 2016

    * Add in form tests from Dancer2::Template::TemplateFlute
    * Add dependency on Dancer2::Template::TemplateFlute

0.200 Thu 26 May 18:28:34 CEST 2016

    * First CPAN release using some code from unreleased Dancer2::Plugin::Form

0.01 Wed Apr  8 12:02:09 2015
    - original version; created by ExtUtils::ModuleMaker 0.52


t/lib/TestApp/TestApp.pm
package TestApp;

use Dancer2;
use Dancer2::Plugin::Form;

post '/init_one_form' => sub {
    my $form_args = from_json(request->params->{form});

    my $form = form('test-form');
    $form->fields( $form_args->{fields} );
    $form->to_session;

    return to_json({ %{$form} });
};

post '/one_form' => sub {
    my $form = form('test-form');
    my $values = $form->values();
    $form->to_session;
    return to_json({ %{$form} });
};

get '/one_form' => sub {    
    my $form = form('test-form');
    $form->values('session');
    return to_json({ %{$form} });
};

get '/reset_test' => sub {
    my $form = form('test-form');
    $form->reset;
    return to_json({ %{$form} });        
};

get '/fill_test' => sub {
    my $form = form('test-form');
    $form->fields(['first-name', 'last-name']);
    $form->fill({
        'first-name' => 'John',
        'last-name' => 'doe'
    });
    $form->to_session;
    return to_json({ %{$form} });
};
get '/fail_test' => sub {
    my $form = form('test-form');
    $form->values('session');
    $form->failure( errors => {
        'last-name' => 'Last name must be capitalized.'
    });
    return to_json({ %{$form} });
};
1;

Makefile.PL
use ExtUtils::MakeMaker;
WriteMakefile(
    NAME           => 'Dancer2::Plugin::TemplateFlute',
    VERSION_FROM   => 'lib/Dancer2/Plugin/TemplateFlute.pm',
    ABSTRACT_FROM  => 'lib/Dancer2/Plugin/TemplateFlute.pm',
    AUTHOR         => 'Evan Brown (evan@bottlenose-wine.com)',
    LICENSE        => 'artistic_2',
    BUILD_REQUIRES => {
        'aliased'       => 0,
        'JSON::MaybeXS' => 0,
        'Test::More'    => '0.94',    # subtest
        'Test::Deep'    => 0,
        'Test::Fatal'   => 0,
    },
    PREREQ_PM => {
        'Carp'             => 0,
        'Dancer2'          => '0.200000',
        'Dancer2::Plugin'  => 0,
        'Dancer2::Template::TemplateFlute' => '0.201',
        'Hash::MultiValue' => '0.12',
        'Moo'              => '2.000000',
        'namespace::clean' => 0,
        'Type::Tiny'       => 0,
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources   => {
            repository => {
                type => 'git',
                url => 'https://github.com/interchange/Dancer2-Plugin-TemplateFlute.git',
                web => 'https://github.com/interchange/Dancer2-Plugin-TemplateFlute',
                bugtracker => 'https://github.com/interchange/Dancer2-Plugin-TemplateFlute/issues',
            },
            IRC => 'irc://irc.perl.org/#dancer',
        },
    },
    dist => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Dancer2-Plugin-TemplateFlute-*' },

);


lib/Dancer2/Plugin/TemplateFlute/Form.pm
package Dancer2::Plugin::TemplateFlute::Form;

use Carp;
use Hash::MultiValue;
use Types::Standard -types;
use Moo;
use namespace::clean;

=head1 NAME

Dancer2::Plugin::TemplateFlute::Form - form object for Template::Flute

=cut

my $_coerce_to_hash_multivalue = sub {
    if ( !defined $_[0] ) {
        Hash::MultiValue->new;
    }
    elsif ( ref( $_[0] ) eq 'Hash::MultiValue' ) {
        $_[0];
    }
    elsif ( ref( $_[0] ) eq 'HASH' ) {
        Hash::MultiValue->from_mixed( $_[0] );
    }
    else {
        croak "Unable to coerce to Hash::MultiValue";
    }
};

#
# attributes
#

has action => (
    is        => 'ro',
    isa       => Str,
    predicate => 1,
    writer    => 'set_action',
);

has errors => (
    is      => 'ro',
    lazy    => 1,
    isa     => InstanceOf ['Hash::MultiValue'],
    default => sub { Hash::MultiValue->new },
    coerce  => $_coerce_to_hash_multivalue,
    clearer => 1,
    writer  => '_set_errors',
);

sub add_error {
    my $self = shift;
    $self->errors->add(@_);
}

sub set_error {
    my $self = shift;
    $self->errors->set(@_);
}

sub set_errors {
    my $self = shift;
    $self->_set_errors(@_);
}

after 'add_error', 'set_error', 'set_errors' => sub {
    $_[0]->set_valid(0);
};

has fields => (
    is      => 'ro',
    lazy    => 1,
    isa     => ArrayRef [Str],
    default => sub { [] },
    clearer => 1,
    writer  => 'set_fields',
);

has log_cb => (
    is        => 'ro',
    isa       => CodeRef,
    predicate => 1,
);

has name => (
    is      => 'ro',
    isa     => Str,
    default => 'main',
);

has pristine => (
    is      => 'ro',
    isa     => Defined & Bool,
    default => 1,
    writer  => 'set_pristine',
);

has session => (
    is       => 'ro',
    isa      => HasMethods [ 'read', 'write' ],
    required => 1,
);

# We use a private writer since we want to have to_session called whenever
# the public set_valid method is called but we also have a need to be
# able to update this attribute without writing the form back to the session.
has valid => (
    is      => 'ro',
    isa     => Bool,
    clearer => 1,
    writer  => '_set_valid',
);

sub set_valid {
    my ( $self, $value ) = @_;
    $self->_set_valid($value);
    $self->log( "debug", "Setting valid for form ",
        $self->name, " to $value." );
    $self->to_session;
}

has values => (
    is      => 'ro',
    lazy    => 1,
    isa     => InstanceOf ['Hash::MultiValue'],
    default => sub { Hash::MultiValue->new },
    coerce  => $_coerce_to_hash_multivalue,
    trigger => sub { $_[0]->set_pristine(0) if $_[1]->keys },
    clearer => 1,
    writer  => 'fill',
);

#
# methods
#

sub errors_hashed {
    my $self = shift;
    my @hashed;

    $self->errors->each(
        sub { push @hashed, +{ name => $_[0], label => $_[1] } } );

    return \@hashed;
}

sub from_session {
    my ($self) = @_;

    $self->log( debug => "Reading form ", $self->name, " from session");

    if ( my $forms_ref = $self->session->read('form') ) {
        if ( exists $forms_ref->{ $self->name } ) {
            my $form = $forms_ref->{ $self->name };

            # set_valid causes write back to session so use private
            # method instead. Also set_errors causes set_valid to be
            # called so use private method there too.
            $self->set_action( $form->{action} ) if $form->{action};
            $self->set_fields( $form->{fields} ) if $form->{fields};
            $self->_set_errors( $form->{errors} ) if $form->{errors};
            $self->fill( $form->{values} )       if $form->{values};
            $self->_set_valid( $form->{valid} ) if defined $form->{valid};

            return 1;
        }
    }
    return 0;
}

sub log {
    my ($self, $level, @message) = @_;
    $self->log_cb->($level, join('',@message)) if $self->has_log_cb;
}

sub reset {
    my $self = shift;
    $self->clear_fields;
    $self->clear_errors;
    $self->clear_values;
    $self->clear_valid;
    $self->set_pristine(1);
    $self->to_session;
}

sub to_session {
    my $self = shift;
    my ($forms_ref);

    $self->log( debug => "Writing form ", $self->name, " to session");

    # get current form information from session
    $forms_ref = $self->session->read('form');

    # update our form
    $forms_ref->{ $self->name } = {
        action => $self->action,
        name   => $self->name,
        fields => $self->fields,
        errors => $self->errors->mixed,
        values => $self->values->mixed,
        valid  => $self->valid,
    };

    # update form information
    $self->session->write( form => $forms_ref );
}

=head1 ATTRIBUTES

=head2 name

The name of the form.

Defaults to 'main',

=head2 action

The form action.

=over

=item writer: set_action

=item predicate: has_action

=back

=head2 errors
    
Errors stored in a L<Hash::MultiValue> object.

Get form errors:

   $errors = $form->errors;

=over

=item writer: set_errors

Set form errors (this will overwrite all existing errors):
    
    $form->set_errors(
        username => 'Minimum 8 characters',
        username => 'Must contain at least one number',
        email    => 'Invalid email address',
    );

=item clearer: clear_errors

=back

B<NOTE:> Avoid using C<< $form->errors->add() >> or C<< $form->errors->set() >>
since doing that means that L</valid> does not automatically get set to C<0>.
Instead use one of L</add_error> or L</set_error> methods.

=head2 fields

Get form fields:

    $fields = $form->fields;

=over

=item writer: set_fields

    $form->set_fields([qw/username email password verify/]);

=item clearer: clear_fields

=back

=head2 log_cb

A code reference that can be used to log things. Signature must be like:

  $log_cb->( $level, $message );

Logging is via L</log> method.

=over

=item predicate: has_log_cb

=back

=head2 pristine

Determines whether a form is pristine or not.

This can be used to fill the form with default values and suppress display
of errors.

A form is pristine until it receives form field input from the request or
out of the session.

=over

=item writer: set_pristine

=back

=head2 session

A session object. Must have methods C<read> and C<write>.

Required.

=head2 valid

Determine whether form values are valid:

    $form->valid();

Return values are 1 (valid), 0 (invalid) or C<undef> (unknown).

=over

=item writer: set_valid

=item clearer: clear_valid

=back

The form status automatically changes to "invalid" when L</errors> is set
or either L</add_errors> or L</set_errors> are called.
    
=head2 values

Get form values as hash reference:

    $values = $form->values;

=over

=item writer: fill

Fill form values:

    $form->fill({username => 'racke', email => 'racke@linuxia.de'});

=item clearer: clear_values

=back

=head1 METHODS

=head2 add_error

Add an error:

    $form->add_error( $key, $value [, $value ... ]);

=head2 errors_hashed

Returns form errors as array reference filled with hash references
for each error.

For example these L</errors>:

    { username => 'Minimum 8 characters',
      email => 'Invalid email address' }

will be returned as:

    [
        { name => 'username', value => 'Minimum 8 characters'  },
        { name => 'email',    value => 'Invalid email address' },
    ]

=head2 from_session

Loads form data from session key C<form>.
Returns 1 if session contains data for this form, 0 otherwise.

=head2 log $level, @message

Log message via L</log_cb>.

=head2 reset

Reset form information (fields, errors, values, valid) and
updates session accordingly.

=head2 set_error

Set a specific error:

    $form->set_error( $key, $value [, $value ... ]);

=head2 to_session

Saves form name, form fields, form values and form errors into 
session key C<form>.


=head1 AUTHORS

Original Dancer plugin by:

Stefan Hornburg (Racke), C<< <racke at linuxia.de> >>

Initial port to Dancer2 by:

Evan Brown (evanernest), C<< <evan at bottlenose-wine.com> >>

Rehacking to Dancer2's plugin2 and general rework:

Peter Mottram (SysPete), C<< <peter at sysnix.com> >>

=head1 BUGS

Please report any bugs or feature requests via GitHub issues:
L<https://github.com/interchange/Dancer2-Plugin-TemplateFlute/issues>.

We will be notified, and then you'll automatically be notified of progress
on your bug as we make changes.

=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2011-2016 Stefan Hornburg (Racke).

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;


lib/Dancer2/Plugin/TemplateFlute.pm
package Dancer2::Plugin::TemplateFlute;

use warnings;
use strict;

use Dancer2::Plugin;
use Dancer2::Plugin::TemplateFlute::Form;

=head1 NAME

Dancer2::Plugin::TemplateFlute - Dancer2 form handler for Template::Flute template engine

=head1 VERSION

Version 0.201

=cut

our $VERSION = '0.201';

plugin_keywords 'form';

sub form {
    my $plugin = shift;

    my $name;
    if ( @_ % 2 ) {
        $name = shift;
    }

    my %params = @_;
    $params{name} = $name if $name;

    $plugin->app->log("debug", "form called with name: $params{name}");

    my $source = delete $params{source};

    # for POST default to body_parameters
    $source = 'body' if ( !$source && $plugin->app->request->is_post );

    if ( $source ) {
        if ( $source eq 'body' ) {
            $params{values} = $plugin->app->request->body_parameters;
        }
        elsif ( $source eq 'query' ) {
            $params{values} = $plugin->app->request->query_parameters;
        }
        elsif ( $source eq 'parameters' ) {
            $params{values} = $plugin->app->request->parameters;
        }
    }

    my $form = Dancer2::Plugin::TemplateFlute::Form->new(
        log_cb  => sub { $plugin->app->logger_engine->log(@_) },
        session => $plugin->app->session,
        %params,
    );

    $form->from_session if ( $source && $source eq 'session' );

    return $form;
}

=head1 SYNOPSIS

Display template with checkout form:
    
    get '/checkout' => sub {
        my $form;

        $form = form( name => 'checkout', source => 'session' );
	
        template 'checkout', { form => $form };
    };

Retrieve form input from checkout form body:

    post '/checkout' => sub {
        my ($form, $values);

        $form = form( name => 'checkout', source => 'body' );
        $values = $form->values;
    };

Reset form after completion to prevent old data from
showing up on new form:

    form('checkout')->reset;

If you have multiple forms then just pass the form token as an array reference:

    get '/cart' => sub {
        my ( $cart_form, $inquiry_form );

        # 'source' defaults to 'session' if not provided
        $cart_form    = form( name => 'cart' );

        # equivalent to form( name => 'inquiry' )
        $inquiry_form = form( 'inquiry' );
	
        template 'checkout', { form => [ $cart_form, $inquiry_form ] };
    };

=head1 KEYWORDS

=head2 form <$name|%params>

The following C<%params> are recognised:

=head3 name

The name of the form. Defaults to 'main'.

=head3 values

    my $form = form( 'main', values => $params );

The form parameters as a L<Hash::MultiValue> object or something that can
be coerced into one.

Instead of L</values> you can also use L</source> to set initial values.

=head3 source

    my $form = form( name => 'main', source => 'body' );

The following values are valid:

=over

=item body

This sets the form values to the request body parameters
L<Dancer2::Core::Request::body_parameters>.

=item query

This sets the form values to the request query parameters
L<Dancer2::Core::Request::query_parameters>.

=item parameters

This sets the form values to the combined body and request query parameters
L<Dancer2::Core::Request::parameters>.

=item session

Reads in values from the session. This is the default if no L</source> or
L</parameters> are specified.

=back

B<NOTE:> if both L</source> and L</values> are supplied then L<values> is
ignored.

See L<Dancer2::Plugin::TemplateFlute::Form> for details of other parameters
that can be used.

=head1 DESCRIPTION
    
C<Dancer2::Plugin::TemplateFlute> is used for forms with the
L<Dancer2::Template::TemplateFlute> templating engine.    

Form fields, values and errors are stored into and loaded from the session key
C<form>.

=head1 AUTHORS

Original Dancer plugin by:

Stefan Hornburg (Racke), C<< <racke at linuxia.de> >>

Initial port to Dancer2 by:

Evan Brown (evanernest), C<< evan at bottlenose-wine.com >>

Rehacking to Dancer2's plugin2 and general rework:

Peter Mottram (SysPete), C<< peter at sysnix.com >>

=head1 BUGS

Please report any bugs or feature requests via GitHub issues:
L<https://github.com/interchange/Dancer2-Plugin-TemplateFlute/issues>.

We will be notified, and then you'll automatically be notified of progress
on your bug as we make changes.

=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2011-2016 Stefan Hornburg (Racke).

Copyright 2015-1016 Evan Brown.

Copyright 2016 Peter Mottram (SysPete).

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;
